src/engine/CartesianProductJoin.cpp

```diff
@@ -131,61 +120,52 @@ void CartesianProductJoin::writeResultColumn(std::span<Id> targetColumn,
     firstInputElementIdx = 0;
   }
 }
+
 // ____________________________________________________________________________
 ProtoResult CartesianProductJoin::computeResult(
     [[maybe_unused]] bool requestLaziness) {
-  IdTable result{getExecutionContext()->getAllocator()};
-  result.setNumColumns(getResultWidth());
-  std::vector<std::shared_ptr<const Result>> subResults;
+  std::vector<std::shared_ptr<const Result>> subResults = calculateSubResults();
 
-  // We don't need to fully materialize the child results if we have a LIMIT
-  // specified and an OFFSET of 0.
-  // TODO<joka921> We could in theory also apply this optimization if a
-  // non-zero OFFSET is specified, but this would make the algorithm more
-  // complicated.
-  std::optional<LimitOffsetClause> limitIfPresent = getLimit();
-  if (!getLimit()._limit.has_value() || getLimit()._offset != 0) {
-    limitIfPresent = std::nullopt;
-  }
-
-  // Get all child results (possibly with limit, see above).
-  for (auto& child : childView()) {
-    if (limitIfPresent.has_value() && child.supportsLimit()) {
-      child.setLimit(limitIfPresent.value());
-    }
-    subResults.push_back(child.getResult());
+  IdTable result = writeAllColumns(subResults);
 
-    const auto& table = subResults.back()->idTable();
-    // Early stopping: If one of the results is empty, we can stop early.
-    if (table.empty()) {
-      break;
-    }
+  // Dereference all the subresult pointers because `getSharedLocalVocabFrom...`
+  // requires a range of references, not pointers.
+  auto subResultsDeref = std::views::transform(
+      subResults, [](auto& x) -> decltype(auto) { return *x; });
```

Comment:
We have a `Functors` or something header where this can be `ad_utility::dereference`.

Reply:
(Note: This will be changed soon anyway).